app/domain/authentication/authn_gcp/restrictions_validator.rb

```diff
@@ -0,0 +1,105 @@
+module Authentication
+  module AuthnGCP
+
+    # This class is responsible of validating resource restrictions combination.
+    # the allowed combination is: Role should have at least 1 permitted constraints, and non of illegal constraints
+    class RestrictionsValidator
+
+      ILLEGAL_STATE_MISSING_CONSTRAINTS_MESSAGE = "constraints are missing or empty"
+      ILLEGAL_STATE_MISSING_RESOURCE_TYPE_MESSAGE = "resource restriction type is missing or empty"
+      ILLEGAL_STATE_DUPLICATE_CONSTRAINTS_MESSAGE = "constraints contain duplications"
+      ILLEGAL_STATE_DUPLICATE_RESTRICTIONS_MESSAGE = "resource restrictions contain duplications"
```

Comment:
Is there a reason why these error messages are not defined as errors in `errors.rb`?

Reply:
I have here a new error which indicate on illegal state that should not happened to the user, but due to the zero trust rule i validate those parameters that are valid before executing the logic.
So i created a general message of illegal state which is similar to all 4 use cases, BYTW i returned here 500 error instead of 401

Reply:
i don't think that this is different from any other configuration error we have. for example, if the user didn't define the `provider-uri` variable in the oidc authenticator policy we return 401. would you return 500 in that case as well?

a 500 response is not very good imo as it `indicates that the server encountered an unexpected condition that prevented it from fulfilling the request`. so it's some kind of "i don't know what happened but i can't fulfill your request". But here we know what happened. specifically, what happened is that the user that is trying to authenticate doesn't have the sufficient info in order to authenticate so i think that a 401 response code will be good.

Reply:
regarding the location of the errors - i still think that they should go in the `errors.rb` class and that we should have 4 different objects, instead of the general `IllegalStateResourceRestrictionsValidation` object that you created.

Reply:
It is not the same example as the `provider-uri` wrong configuration, if those errors (illegal state errors) will happened it is not the user fault.
For example if we will have an error in the DB which will return a nil annotation name or duplicate values (maybe as a result of a bug in a previous conjur upgrade), so it will create a restriction list (which is the input of this class) with un processable values, and to an error that it is a defiantly not 401 error

Reply:
so i don't understand the case. can you describe a use case in which this will happen? for example, how will you test it?

Reply:
It can be tested just with UT

Reply:
i don't think we need to make up issues that can be reproduced